### ISSUE TITLE
refactor(http): rename URLSearchParams to something else

### ISSUE THREAD

**Issue by caitp** [angular/angular#3513]
TL;DR: Rename `URLSearchParams` to `QueryParams`. 

----

I've discussed with @jeffbcross that it would be wise to rename the class, as we are considering having a behaviour which differs from the WHATWG's version of the interface --- naming this differently could help minimize confusion when using the API.

Some ideas:

-`SearchQuery`
-`SearchParameters`
-`SearchParams`

All missing the leading `URL` and maybe standing out from the family of interfaces derived from https://url.spec.whatwg.org.

---

The behaviours discussed, which some may agree or disagree with, include the following:

- An explicit `merge` method (which I think would be valuable to propose to the editors of [whatwg/url](https://github.com/whatwg/url). If the WHATWG did in the future accept a proposal like this, it may be possible to deprecate and remove the in-angular URLSearchParams and rely on the native implementation instead, in the future. So that's something!

Semantic issue: Tab Atkins Jr isn't keen on the fact that sometimes, you want to merge by appending to existing values (a MultiMap behaviour) and sometimes you want to replace existing values (a Map behaviour) --- but doesn't believe there's value in providing a way to specify a schema. Having two methods would certainly require being explicit, though.

OR

- Implicitly merging during RequestOptions.merge(), rather than the WHATWG's strategy of replacing the original query entirely (currently, but possibly not in the future, for all URLs which are associated with the URLSearchParams object). It should be possible to implicitly merge this way even if the WHATWG's version of the object were used, but may not be.

Any thoughts or suggestions are welcome

**Comment by caitp** [comment]
This is really a discussion, because it's not super clear the direction it should go in, and eventually the API needs to be frozen, or become as close to frozen as possible. I invite people to input their perspective on it. If a decision is arrived at quickly, then it will surely be a very short change to make.

**Comment by mhevery** [comment]
@jeffbcross and @btford could we get a resolution, since we need to get all breaking changes in soon.

**Comment by btford** [comment]
I'm not familiar with `URLSearchParams`. Is this for HTTP? If this is for serializing data to send to some end-point, then I think it's a totally distinct issue from how routing interacts with the URL.

**Comment by robwormald** [comment]
it seems like the preferred nomenclature is `query` over `search`, so `QueryParams`? URLParams make me think of `foo/:bar` and Search isn't really a thing.

**Comment by jeffbcross** [comment]
@btford yeah, it's just for Http and is used for serializing parameters. I originally created it as a polyfill for the [`URLSearchParams`](https://url.spec.whatwg.org/#urlsearchparams) constructor that accompanied the [fetch spec](https://fetch.spec.whatwg.org). @caitp had discussed adding/changing some of its merging semantics, and decided that we should probably change its name so as to not be considered a polyfill.

I like @robwormald's suggestion of `QueryParams`. I'll update this issue with QueryParams as the resolution.

Note: I don't think we should change the names of `Headers`, `Response`, or `Request`, even though those conflict with fetch constructor names, and even though they somewhat resemble the constructors as defined by fetch. Those names are generic enough, unlike `URLSearchParams`.

**Comment by jeffbcross** [comment]
Regarding how to handle merging behavior, I'm going to give it some more thought, but I would like to have specialized merging functions not be part of the `QueryParams` prototype to better embrace immutability. I'll open an issue or two on Monday to implement params merging, as well as necessary changes to `RequestOptions` merging.

**Comment by robwormald** [comment]
This issue was moved to angular/http#68
